libs/@local/hash-subgraph/src/types/subgraph.ts

```diff
@@ -1,42 +1,84 @@
-import { GraphResolveDepths } from "@local/hash-graph-client";
+import {
+  type DataTypeRootType as DataTypeRootTypeBp,
+  type EntityRootType as EntityRootTypeBp,
+  type EntityTypeRootType as EntityTypeRootTypeBp,
+  type PropertyTypeRootType as PropertyTypeRootTypeBp,
+  type SubgraphRootType as SubgraphRootTypeBp,
+} from "@blockprotocol/graph";
+import { Subtype } from "@local/advanced-types/subtype";
 
-import { Edges } from "./edge";
 import {
   DataTypeWithMetadata,
   Entity,
   EntityTypeWithMetadata,
   PropertyTypeWithMetadata,
 } from "./element";
-import { EntityVertexId, OntologyTypeVertexId } from "./identifier";
-import { ResolvedTimeProjection, TimeProjection } from "./time";
-import { Vertices } from "./vertex";
+import { Edges } from "./subgraph/edges";
+import { GraphResolveDepths } from "./subgraph/graph-resolve-depths";
+import { SubgraphTemporalAxes } from "./subgraph/temporal-axes";
+import {
+  EntityVertexId,
+  OntologyTypeVertexId,
+  Vertices,
+} from "./subgraph/vertices";
 
-export type SubgraphRootTypes = {
-  dataType: {
+export * from "./subgraph/edges";
+export * from "./subgraph/graph-resolve-depths";
+export * from "./subgraph/temporal-axes";
+export * from "./subgraph/vertices";
+
+export type DataTypeRootType = Subtype<
+  DataTypeRootTypeBp,
+  {
     vertexId: OntologyTypeVertexId;
     element: DataTypeWithMetadata;
-  };
-  propertyType: {
+  }
+>;
+
+export type PropertyTypeRootType = Subtype<
+  PropertyTypeRootTypeBp,
+  {
     vertexId: OntologyTypeVertexId;
     element: PropertyTypeWithMetadata;
-  };
-  entityType: {
+  }
+>;
+
+export type EntityTypeRootType = Subtype<
+  EntityTypeRootTypeBp,
+  {
     vertexId: OntologyTypeVertexId;
     element: EntityTypeWithMetadata;
-  };
-  entity: {
+  }
+>;
+
+export type EntityRootType = Subtype<
+  EntityRootTypeBp<true>,
+  {
     vertexId: EntityVertexId;
     element: Entity;
-  };
-};
+  }
+>;
+
+export type SubgraphRootType = Subtype<
+  SubgraphRootTypeBp<true>,
+  DataTypeRootType | PropertyTypeRootType | EntityTypeRootType | EntityRootType
+>;
 
-export type SubgraphRootType = SubgraphRootTypes[keyof SubgraphRootTypes];
+/** @todo - Figure out the incompatible vertices/edges is it the `&` instead of `|`? */
+// export type Subgraph<
+//   RootType extends SubgraphRootType = SubgraphRootType,
+//   > = Subtype<{
+//   roots: RootType["vertexId"][];
+//   vertices: Vertices;
+//   edges: Edges;
+//   depths: GraphResolveDepths;
+//   temporalAxes: SubgraphTemporalAxes
+// }, SubgraphBp<true, RootType>;
```

Comment:
I'm really not sure what to do about this. I think it's to do with the index types on the inner objects (`Vertices` and `Edges`). The changes we've made here give us what is, in many ways, a better type than the upstream Block Protocol one, as the index types to the inner objects (e.g. `EntityId` or `BaseUri`) are all branded, and do not overlap. Because of this TS is able to actually know what thing we're referring to, so if you do `Vertices[BaseUri]`, it knows the values inside can only be knowledge graph vertices, whereas in the BP those types completely overlap (they are just `string`s), which means that you have to tell it what the inner type is, even when you know you're dealing with an `entityId`.

Because of this, I think the types actually _aren't_ compatible as the BP one is more permissive than we intend it to be (and the spec/definition we associate with a `Subgraph` will probably be more constrained than the type), and thus it's actually somewhat incorrect and **won't** be assignable.

This causes annoyances elsewhere (it means anything using a `Subgraph` isn't assignable either, such as the Graph callback definitions), and means we have to do `as unknown as Subgraph..` casts which isn't particularllyy safe. (We could create `mapLocalSubgraphToBp` and `mapSubgraphBpToLocal` functions I guess, and how in-depth we decide to make those beyond `as` casts would need to have considerations made about performance).